patches/server/0852-Multi-Block-Change-API-Implementation.patch

```diff
@@ -0,0 +1,79 @@
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+@@ -30,6 +30,7 @@ import javax.annotation.Nullable;
+ import net.minecraft.Util;
+ import net.minecraft.advancements.AdvancementProgress;
+ import net.minecraft.core.BlockPos;
++import net.minecraft.core.SectionPos; // Paper
+ import net.minecraft.nbt.CompoundTag;
+ import net.minecraft.network.FriendlyByteBuf;
+ import net.minecraft.network.chat.ChatType;
+@@ -46,6 +47,7 @@ import net.minecraft.network.protocol.game.ClientboundLevelEventPacket;
+ import net.minecraft.network.protocol.game.ClientboundLevelParticlesPacket;
+ import net.minecraft.network.protocol.game.ClientboundMapItemDataPacket;
+ import net.minecraft.network.protocol.game.ClientboundPlayerInfoPacket;
++import net.minecraft.network.protocol.game.ClientboundSectionBlocksUpdatePacket; // Paper
+ import net.minecraft.network.protocol.game.ClientboundSetDefaultSpawnPositionPacket;
+ import net.minecraft.network.protocol.game.ClientboundSetEquipmentPacket;
+ import net.minecraft.network.protocol.game.ClientboundSetExperiencePacket;
+@@ -74,6 +76,7 @@ import net.minecraft.world.level.GameType;
+ import net.minecraft.world.level.block.Blocks;
+ import net.minecraft.world.level.biome.BiomeManager;
+ import net.minecraft.world.level.block.entity.SignBlockEntity;
++import net.minecraft.world.level.block.state.BlockState; // Paper
+ import net.minecraft.world.level.saveddata.maps.MapDecoration;
+ import net.minecraft.world.level.saveddata.maps.MapItemSavedData;
+ import net.minecraft.world.phys.Vec3;
+@@ -836,6 +839,44 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+         this.getHandle().connection.send(packet);
+     }
+ 
++    // Paper start
++    @Override
++    public void sendMultiBlockChange(Map<Location, BlockData> blockChanges, boolean suppressLightUpdates) {
++        if (this.getHandle().connection == null) return;
++
++        Map<SectionPos, Map<Short, BlockState>> sectionMap = new HashMap<>();
++
++        for (Map.Entry<Location, BlockData> entry : blockChanges.entrySet()) {
++            Location location = entry.getKey();
++            if (!location.getWorld().equals(this.getWorld())) continue;
++
++            BlockData blockData = entry.getValue();
++            BlockPos blockPos = new BlockPos(location.getBlockX(), location.getBlockY(), location.getBlockZ());
++            SectionPos sectionPos = SectionPos.of(blockPos);
++
++            Map<Short, BlockState> sectionData = sectionMap.computeIfAbsent(sectionPos, key -> new HashMap<>());
++            sectionData.put(SectionPos.sectionRelativePos(blockPos), ((CraftBlockData) blockData).getState());
++        }
++
++        for (Map.Entry<SectionPos, Map<Short, BlockState>> entry : sectionMap.entrySet()) {
++            SectionPos sectionPos = entry.getKey();
++            Map<Short, BlockState> blockData = entry.getValue();
++
++            FriendlyByteBuf friendlyByteBuf = new FriendlyByteBuf(Unpooled.buffer());
```

Comment:
Manual writing is a nono here - either make a new ClientboundSectionBlocksUpdatePacket constructor or temporarily modify the `states` block container in LevelChunkSection

Reply:
I went ahead and made a new constructor in latest changes, wasn't sure if I should try and use what was available to do it or just make a new constructor initially.